consensus/scp/tests/mock_network/mod.rs

```diff
@@ -414,41 +338,63 @@ impl SCPNode {
                                 SCPNodeTaskMessage::StopTrigger => {
                                     break 'main_loop;
                                 }
-                            };
+                            },
+                            Err(_) => {
+                                // Yield to other threads when we don't get a new message
+                                // This improves performance significantly.
+                                std::thread::yield_now();
+                            }
+                        };
+
+                        let incoming_msgs_count = incoming_msgs.len();
+
+                        if !(incoming_msgs_count == 0 || incoming_msgs_count == 1) {
+                            log::error!(logger, "incoming_msgs_count > 1");
+                            assert!(incoming_msgs_count == 0 || incoming_msgs_count == 1);
+                            // exit
                         }
 
                         // Process values submitted to our node
                         if !pending_values.is_empty() {
-                            let vals = pending_values
-                                              .iter()
-                                              .cloned()
-                                              .collect::<Vec<String>>();
+                            let mut vals = pending_values.iter().cloned().collect::<Vec<String>>();
 
-                            let outgoing_msg : Option<Msg<String>> = {
-                                thread_local_node
-                                .lock()
-                                .expect("lock failed on node nominating value")
-                                .nominate(
-                                    current_slot as SlotIndex,
-                                    BTreeSet::from_iter(vals),
-                                )
-                                .expect("node.nominate() failed")
-                            };
+                            if allow_propose {
+                                if vals.len() > test_options.max_values_per_slot {
+                                    vals.sort();
+                                    vals.truncate(test_options.max_values_per_slot);
+                                }
+                                allow_propose = false;
+                            } else {
+                                vals.clear();
+                            }
 
-                            if let Some(outgoing_msg) = outgoing_msg {
-                                (broadcast_msg_fn)(logger.clone(), outgoing_msg);
-                                total_broadcasts += 1;
+                            if vals.len() > 0 {
```

Comment:
As mentioned in the previous PR, we use `is_empty()` in order places in the codebase.